### PR TITLE
Sweepers: Allow skipping individual IAM Roles

### PR DESCRIPTION
**Pull request by gdavison**
Skips individual IAM Role resources during sweeping on `AccessDenied` error.

Relates #16570

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
NONE
```

Output from acceptance testing:

```console
$ make testacc TESTARGS='-run=TestAccAWSIAMRole_'

--- PASS: TestAccAWSIAMRole_badJSON (88.70s)
--- PASS: TestAccAWSIAMRole_disappears (183.12s)
--- PASS: TestAccAWSIAMRole_basic (233.11s)
--- PASS: TestAccAWSIAMRole_force_detach_policies (236.80s)
--- PASS: TestAccAWSIAMRole_namePrefix (257.41s)
--- PASS: TestAccAWSIAMRole_tags (309.68s)
--- PASS: TestAccAWSIAMRole_testNameChange (314.97s)
--- PASS: TestAccAWSIAMRole_MaxSessionDuration (344.53s)
--- PASS: TestAccAWSIAMRole_basicWithDescription (357.99s)
--- PASS: TestAccAWSIAMRole_PermissionsBoundary (426.20s)
```

```console
$ make sweep SWEEPARGS='-sweep-run=aws_iam_role'

2020/12/03 13:02:29 [DEBUG] Running Sweeper (aws_iam_role) in region (us-west-2)
2020/12/03 13:02:29 [DEBUG] Deleting IAM Role (tf_acctest_role1)
2020/12/03 13:02:31 [WARN] Skipping IAM Role (tf_acctest_role1): unable to detach policies: AccessDenied: User: arn:aws:iam::123456789012:user/jdoe is not authorized to perform: iam:DetachRolePolicy on resource: role tf_acctest_role1 with an explicit deny
	status code: 403, request id: 
2020/12/03 13:02:31 [DEBUG] Deleting IAM Role (tf_acctest_role2)
2020/12/03 13:02:32 [WARN] Skipping IAM Role (tf_acctest_role2): unable to detach policies: AccessDenied: User: arn:aws:iam::123456789012:user/jdoe is not authorized to perform: iam:DetachRolePolicy on resource: role tf_acctest_role2 with an explicit deny
	status code: 403, request id: 
2020/12/03 13:02:32 [DEBUG] Deleting IAM Role (tf_acctest_role3)
2020/12/03 13:02:33 [WARN] Skipping IAM Role (tf_acctest_role3): unable to detach policies: AccessDenied: User: arn:aws:iam::123456789012:user/jdoe is not authorized to perform: iam:DetachRolePolicy on resource: role tf_acctest_role3 with an explicit deny
	status code: 403, request id: 
2020/12/03 13:02:33 Sweeper Tests ran successfully:
	...
	- aws_iam_role
```

### CODE REVIEW

File: aws/resource_aws_iam_role.go [hunk 1]
```diff
@@ -347,30 +356,21 @@ func resourceAwsIamRoleDelete(d *schema.ResourceData, meta interface{}) error {
...
 		RoleName: aws.String(rolename),
 	})
-
-	if isAWSErr(err, iam.ErrCodeNoSuchEntityException, "") {
+	if tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
 		return nil
 	}
-
 	if err != nil {
-		return fmt.Errorf("Error listing Profiles for IAM Role (%s) when trying to delete: %s", rolename, err)
```
Comment (anGie44) [inline aws/resource_aws_iam_role.go:373]: just curious if we'll be missing this fine-grain info if we're just returning the `err` or is the API generally pretty good about returning the error with enough info for practitioners to debug?

File: aws/resource_aws_iam_role.go [hunk 2]
```diff
@@ -311,34 +313,41 @@ func resourceAwsIamRoleUpdate(d *schema.ResourceData, meta interface{}) error {
 }
 
 func resourceAwsIamRoleDelete(d *schema.ResourceData, meta interface{}) error {
-	iamconn := meta.(*AWSClient).iamconn
+	conn := meta.(*AWSClient).iamconn
 
-	// Roles cannot be destroyed when attached to an existing Instance Profile
```
Comment (anGie44) [inline aws/resource_aws_iam_role.go:316]: maybe keep the removed comments so we don't lose the rational if this Delete function needs to be modified in the future?

**Review by anGie44 — APPROVED** [review]
Just a couple minor comments around the inline comments/err messaging after rebasing, otherwise looks great 👍

File: aws/resource_aws_iam_role.go (hunk 1 above)
Comment (gdavison) [reply aws/resource_aws_iam_role.go:373]: This error case was several levels of wrapping, so we'd get something noisy like

> error deleting IAM Role (\<rolename>) instance profiles: Error listing Profiles for IAM Role (\<rolename>) when trying to delete: \<error>

File: aws/resource_aws_iam_role.go (hunk 2 above)
Comment (gdavison) [reply aws/resource_aws_iam_role.go:316]: Good point

**Pull request merged**